In [1]:
!pip install snorkel

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.3/103.3 kB 6.4 MB/s eta 0:00:00


In [2]:
import pandas as pd
from google.colab import files

First, inspect the available labeled data to notice reccurring patterns or words specific to a label

In [3]:
files.upload()

Saving Food-Health-Claims-KG_clean.csv to Food-Health-Claims-KG_clean.csv


{'Food-Health-Claims-KG_clean.csv': b'"Status","Claim","Claim Type","Food","NEW Food Type","Food Type","Food Matrix OLD","Regulation specifications","NEW Food Matrix","Food Ontology Term","Food Amount OLD","NEW Food Amount","Unit","Value","Frequency","Target population","Target population ontology term","Extra information","Phenotype","Phenotype Ontology Term","Phenotype Onto Term Match","Relationship-effect","Health relationship","EFSA Opinion Reference","Supporting Evidence Text 1","Supporting Evidence Reference 1","Supporting Evidence Text 2","Supporting Evidence Reference 2","Supporting Evidence Text 3","Supporting Evidence Reference 3","Supporting Evidence Text 4","Supporting Evidence Reference 4","Supporting Evidence Text 5","Supporting Evidence Reference","Supporting Evidence Text 5","Supporting Evidence Reference 5","Supporting Evidence Text 6","Supporting Evidence Reference 6","Supporting Evidence Text 7","Supporting Evidence Reference 7","Supporting Evidence Text 8","Supporti

#Inspect labeled data for patterns

In [4]:
data = pd.read_csv("/content/Food-Health-Claims-KG_clean.csv")

In [5]:
def printrow(row):
    claim = row["Claim"].lower()
    food = row["Food"].lower()
    pheno = row["Phenotype"].lower()
    rel = row["Relationship-effect"].lower()
    print(claim,' : ',  food,' : ',pheno )
    print(rel)

In [6]:
enh = data[data["Relationship-effect"] == "Enhancing a function"]
red = data[data["Relationship-effect"] == "Reducing a risk factor"]
main = data[data["Relationship-effect"] == "Maintenance of a function"]
enh

Status                                              Claim  Claim Type  \
2    Finished  Activated charcoal contributes to reducing exe...  Art. 13(1)   
3    Finished  Barley grain fibre contributes to an increase ...  Art. 13(1)   
22   Finished  Carbohydrate-electrolyte solutions enhance the...  Art. 13(1)   
47   Finished  Creatine increases the physical performance in...  Art. 13(1)   
56   Finished  Folate contributes to maternal tissue growth d...  Art. 13(1)   
62   Finished  Folate contributes to the reduction of tiredne...  Art. 13(1)   
78   Finished  Iron contributes to the reduction of tiredness...  Art. 13(1)   
80   Finished  Lactase enzyme improves lactose digestion in i...  Art. 13(1)   
81   Finished  Lactulose contributes to an acceleration of in...  Art. 13(1)   
83   Finished  Live cultures in yoghurt or fermented milk imp...  Art. 13(1)   
84   Finished  Magnesium contributes to a reduction of tiredn...  Art. 13(1)   
98   Finished  Meat or fish contributes to the improvement of...  Art. 13(1)   
99   Finished  Melatonin contributes to the allevation of sub...  Art. 13(1)   
100  Finished  Melatonin contributes to the reduction of time...  Art. 13(1)   
108  Finished  Niacin contributes to the reduction of tiredne...  Art. 13(1)   
109  Finished  Oat grain fibre contributes to an increase in ...  Art. 13(1)   
114  Finished  Pantothenic acid contributes to the reduction ...  Art. 13(1)   
124  Finished     Protein contributes to a growth in muscle mass  Art. 13(1)   
140  Finished  Riboflavin contributes to the reduction of tir...  Art. 13(1)   
170  Finished  Vitamin B12 contributes to the reduction of ti...  Art. 13(1)   
180  Finished  Vitamin B6 contributes to the reduction of tir...  Art. 13(1)   
194  Finished  Vitamin C contributes to the reduction of tire...  Art. 13(1)   
196  Finished                Vitamin C increases iron absorption  Art. 13(1)   
207  Finished  Walnuts contribute to the improvement of the e...  Art. 13(1)   
210  Finished  Wheat bran fibre contributes to an acceleratio...  Art. 13(1)   
211  Finished  Wheat bran fibre contributes to an increase in...  Art. 13(1)   
231  Finished  Carbohydrates contribute to the recovery of no...  Art. 13(5)   
234  Finished  Daily creatine consumption can enhance the eff...  Art. 13(5)   
236  Finished  Sugar beet fibre contributes to an increase in...  Art. 13(5)   

                                   Food NEW Food Type          Food Type  \
2                    Activated charcoal      Compound  BioactiveCompound   
3                    Barley grain fibre      Nutrient      Macronutrient   
22   Carbohydrate-electrolyte solutions          Food               Food   
47                             Creatine      Nutrient      Macronutrient   
56                               Folate      Nutrient      Micronutrient   
62                               Folate      Nutrient      Micronutrient   
78                                 Iron      Nutrient      Micronutrient   
80                       Lactase enzyme      Compound          Substance   
81                            Lactulose      Nutrient      Macronutrient   
83                Live yoghurt cultures      Compound          Substance   
84                            Magnesium      Nutrient      Micronutrient   
98                         Meat or fish          Food               Food   
99                            Melatonin      Compound          Substance   
100                           Melatonin      Compound          Substance   
108                              Niacin      Nutrient      Micronutrient   
109                     Oat grain fibre      Nutrient      Macronutrient   
114                    Pantothenic acid      Nutrient      Micronutrient   
124                             Protein      Nutrient      Macronutrient   
140             Riboflavin (Vitamin B2)      Nutrient      Micronutrient   
170                         Vitamin B12      Nutrient      Micronutrient   
180       

In [9]:
for i in range(29):
  printrow(enh.iloc[i])
  print()

activated charcoal contributes to reducing exessive flatulence after eating  :  activated charcoal  :  excessive flatulence
enhancing a function

barley grain fibre contributes to an increase in faecal bulk  :  barley grain fibre  :  increase in faecal bulk
enhancing a function

carbohydrate-electrolyte solutions enhance the absorption of water during physical exercise  :  carbohydrate-electrolyte solutions  :  enhancement of water absorption during physical exercise
enhancing a function

creatine increases the physical performance in successive bursts of short-term, high intensity exercise   :  creatine  :  increased physical performance
enhancing a function

folate contributes to maternal tissue growth during pregnancy  :  folate  :  maternal tissue growth 
enhancing a function

folate contributes to the reduction of tiredness and fatigue  :  folate  :  reduction of tiredness and fatigue
enhancing a function

iron contributes to the reduction of tiredness and fatigue  :  iron  :  dec

Looking at the claims above, we notice that the words "enhance", "improve", "increase" and "acceleration" and their variants are recurring within the "enhancing a function" label.


In [8]:
for i in range(23):
  printrow(red.iloc[i])
  print()

consumption of hydroxypropyl methylcellulose with a meal contributes to a reduction in the blood glucose rise after that meal   :  hydroxypropyl methylcellulose (hpmc)  :  increased blood glucose
reducing a risk factor

consumption of alpha-cyclodextrin as a part of a starch-containing meal contributes to the reduction of the blood glucose rise after that meal  :  alpha-cyclodextrin  :  increased blood glucose
reducing a risk factor

consumption of arabinoxylan as part of a meal contributes to a reduction of the blood glucose rise after that meal   :  arabinoxylan produced from wheat endosperm  :  increased blood glucose
reducing a risk factor

consumption of beta-glucans from oats or barley as part of a meal contributes to the reduction of the blood glucose rise after that meal   :  beta-glucans from oats and barley  :  increased blood glucose
reducing a risk factor

consumption of foods containing fructose leads to a lower blood glucose rise compared to foods containing sucrose or gl

Looking at the claims above, we notice that the words : "reduction", "lower", "protection", "neutralisation" and their variants are often associated with the label "reducing a risk factor"

In [10]:
for i in range(90,120):
  printrow(main.iloc[i])
  print()

phosphorus contributes to the maintenance of normal teeth  :  phosphorus  :  normal teeth
maintenance of a function

plant sterols/stanols contribute to the maintenance of normal blood cholesterol levels  :  plant sterols and plant stanols  :  normal blood cholesterol
maintenance of a function

potassium contributes to normal functioning of the nervous system  :  potassium  :  normal functioning of the nervous system
maintenance of a function

potassium contributes to normal muscle function  :  potassium  :  normal muscle function
maintenance of a function

potassium contributes to the maintenance of normal blood pressure  :  potassium  :  blood pressure normal
maintenance of a function

protein contributes to the maintenance of muscle mass  :  protein  :  muscle mass
maintenance of a function

protein contributes to the maintenance of normal bones  :  protein  :  normal bone
maintenance of a function

reducing consumption of saturated fat contributes to the maintenance of normal blood

Looking at the claims above, we notice that the words "maintenance", "normal" appear often and can be associated with the label "maintenance of a function"

# Apply the patterns recognized to the unlabeled data using Snorkel

In [11]:
#Import unlabeled data
files.upload()

Saving EU_Register_on_nutrition_and_health_claims.csv to EU_Register_on_nutrition_and_health_claims.csv


{'EU_Register_on_nutrition_and_health_claims.csv': b'"Claim type","Nutrient substance, food or food category"," Claim"," Conditions of use of the claim / Restrictions of use / Reasons for non-authorisation"," Health relationship"," EFSA opinion reference"," Commission Regulation"," Status"," Entry Id"\n"Art.13(1)","Alpha-linolenic acid (ALA)","ALA contributes to the maintenance of normal blood cholesterol levels","The claim may be used only for food which is at least a source of ALA as referred to in the claim SOURCE OF OMEGA 3 FATTY ACIDS as listed in the Annex to Regulation (EC) No 1924/2006. \nInformation shall be given to the consumer that the beneficial effect is obtained with a daily intake of 2 g of ALA.","maintenance of normal blood cholesterol concentrations","2009;7(9):1252","Commission Regulation (EU) 432/2012 of 16/05/2012","Authorised","493, 568"\n"","","","","","2011;9(6):2203","","",""\n"Art.13(1)","Activated charcoal","Activated charcoal contributes to reducing excessiv

In [37]:
df = pd.read_csv("/content/EU_Register_on_nutrition_and_health_claims.csv")
df

Claim type          Nutrient substance, food or food category  \
0        Art.13(1)                         Alpha-linolenic acid (ALA)   
1              NaN                                                NaN   
2        Art.13(1)                                 Activated charcoal   
3        Art.13(1)                                 Barley grain fibre   
4        Art.13(1)                                       Beta-glucans   
...            ...                                                ...   
2558  Art.14(1)(b)  Mixture of bifidobacteria (Bifidobacterium bif...   
2559  Art.14(1)(b)  I omega kids®/Pufan 3 kids® containing docosah...   
2560  Art.14(1)(b)          Dairy products (milk, cheese and yoghurt)   
2561  Art.14(1)(b)  regulat®.pro.kid IMMUN, food supplement with f...   
2562  Art.14(1)(b)  regulat®.pro.kid.BRAIN, food supplement with p...   

                                                  Claim  \
0     ALA contributes to the maintenance of normal b...   
1                                                   NaN   
2     Activated charcoal contributes to reducing exc...   
3     Barley grain fibre contributes to an increase ...   
4     Beta-glucans contribute to the maintenance of ...   
...                                                 ...   
2558  Probiotic bifidobacteria lead to a healthy int...   
2559  Provide serenity and room for a beneficial dev...   
2560  Three portions of dairy food everyday, as part...   
2561  regulat®.pro.kid IMMUN supports, stimulates an...   
2562  regulat®.pro.kid.BRAIN contributes to mental a...   

      Conditions of use of the claim / Restrictions of use / Reasons for non-authorisation  \
0     The claim may be used only for food which is a...                                      
1                                                   NaN                                      
2     The claim may be used only for food which cont...                                      
3     The claim may be used only for food which is h...                                      
4     The claim may be used only for food which cont...                                      
...                                                 ...                                      
2558  Non-compliance with the Regulation because on ...                                      
2559  Non-compliance with the Regulation because on ...                                      
2560  Non-compliance with the Regulation because on ...                                      
2561  Non-compliance with the Regulation because on ...                                      
2562  Non-compliance with the Regulation because on ...                                      

                                    Health relationship  \
0     maintenance of normal blood cholesterol concen...   
1                                                   NaN   
2     reduction of excessive intestinal gas accumula...   
3                               increase in faecal bulk   
4     maintenance of normal blood cholesterol concen...   
...                                                 ...   
2558                                                NaN   
2559                                                NaN   
2560                                                NaN   
2561                                                NaN   
2562                                                NaN   

      EFSA opinion reference  \
0             2009;7(9):1252   
1             2011;9(6):2203   
2             2011;9(4):2049   
3             2011;9(6):2249   
4             2009;7(9):1254   
...                      ...   
2558            Q-2009-00224   
2559   Q-2008-092 Q-2008-097   
2560              Q-2008-110   
2561              Q-2008-082   
2562              Q-2008-083   

                                  Commission Regulation          Status  \
0     Commission Regulation (EU) 432/2012 of 16/05/2012      Authorised   
1                                                

In [38]:
ddf = df[df[" Status"] == "Non-authorised" ]
ddf = ddf.reset_index()
data = pd.DataFrame()
data["text"] = ddf[" Claim"]
data["Food"] = ddf["Nutrient substance, food or food category"]
data["Phenotype"] = ddf[" Health relationship"]

In [39]:
#Inspect data
mat = data["Phenotype"].isna()
for i in range(len(mat)):
    if mat.iloc[i]:
        print(i)

1875
1876
1877
1878
1879
1880
1881
1882
1883
1884
1885
1886
1887
1888
1889
1890
1891
1892
1893
1894
1895
1896
1897
1898
1899
1900
1901
1902
1903
1904
1905
1906
1907
1908
1909
1910
1911
1912
1913
1914
1915
1916
1917
1918
1919
1920
1921
1922
1923
1924
1925
1926
1927
1928
1929
1930
1931
1932
1933
1934
1935
1936
1937
1938
1939
1940
1941
1942
1943
1944
1945
1946
1947
1948
1949
1950
1951
1952
1953
1954
1955
1956
1957
1958
1959
1960
1961
1962
1963
1964
1965
1966
1967
1968
1969
1970
1971
1972
1973
1974
1975
1976
1977
1978
1979
1980
1981
1982
1983
1984
1985
1986
1987
1988
1989
1990
1991
1992
1993
1994
1995
1996
1997
1998
1999
2000
2001
2002
2003
2004
2005
2006
2007
2008
2009
2010
2011
2012
2013
2014
2015
2016
2017
2018
2019
2020
2021
2022
2023
2024
2025
2026
2027
2028
2029
2030
2031
2032
2033
2034
2035
2036
2037
2038
2039
2040
2041
2042
2043
2044
2045
2046
2047
2048
2049
2050
2051
2052
2053
2054
2055
2056
2057
2058
2059
2060
2061
2062
2063
2064
2065
2066
2067
2068
2069
2070
2071
2072
2073
2074


In [40]:
#Remove NaN and "not validated" values
data = data[:1875]
aa = data["Phenotype"] == "not validated"
count = 0
for i in range(len(data["text"])):
    if data["Phenotype"].iloc[i] == "not validated":
        print(i)

11
12
13
14
15
16
17
18
19
20
21
22
24
25
29
32
48
49
50
51
52
53
54
55
56
57
58
59
60
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
131
132
133
156
157
175
184
185
186
187
190
191
192
193
195
196
197
198
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
218
219
220
222
223
224
225
226
227
228
229
230
231
234
235
236
237
244
245
246
248
270
275
276
279
284
297
299
300
303
304
305
310
311
312
315
318
319
320
323
346
347
350
351
352
353
354
355
357
358
359
360
361
363
364
365
367
389
393
410
414
427
445
457
458
459
460
461
462
463
464
465
470
471
472
491
513
538
539
540
545
547
548
549
550
551
552
567
573
577
578
586
588
597
666
667
670
673
679
681
685
686
687
692
693
703
704
714
715
716
717
718
719
720
724
726
727
738
739
742
756
757
793
796
798
799
800
801
802
804
809
810
822
825
855
887
892
903
914
916
917
924
941
949
960
962
986
990
992
1001
1006
1007
1008
1014
1015
1022
1023
1024
1029
1030
1032
1034
1044
1045
1046
1047
1048
1051
1055
1056
1057
10

In [41]:
data = data[data["Phenotype"] != "not validated"]
len(data)

1491

In [42]:
data

text  \
0     "- The combination of soluble / insoluble fibr...   
1     "The inclusion of 3 grams oat beta glucan solu...   
2     "Traditionally used for the good circulation o...   
3     "Traditionally used to enhance the renal elimi...   
4     "Traditionally used to facilitate the digestio...   
...                                                 ...   
1860  “Bifidobacterium longum BB536 is a probiotic t...   
1862  “HPMC promotes good digestive health and regul...   
1863  “MK-7 is a highly bioavailable form of vitamin...   
1864  “Provides antioxidant protection.”\n“For daily...   
1874  • For people with a sensitive stomach;\n• Soot...   

                                                   Food  \
0                                      Sugar beet fibre   
1                         CalorieControl Trim® Oat Bran   
2                Citrus sinensis - common name : Orange   
3            Prunus cerasus - common name : Sour cherry   
4            Prunus cerasus - common name : Sour cherry   
...                                                 ...   
1860     Morinaga <em>Bifidobacterium longum</em> BB536   
1862  Hydroxypropyl methylcellulose (HPMC)\nDietary ...   
1863         Menaquinone-7 (MK-7, a form of vitamin K2)   
1864       Chlorophyllin\n[Sodium copper chlorophyllin]   
1874  Fat-reduced cream powder (rich source of milk ...   

                                              Phenotype  
0                             Changes in bowel function  
1     Maintenance or achievement of normal blood glu...  
2                            “Vascular and vein health”  
3                   Increase in renal water elimination  
4                                  “Digestive function”  
...                                                 ...  
1860  decreasing potentially pathogenic intestinal m...  
1862               maintenance of normal bowel function  
1863            function of the heart and blood vessels  
1864          reduction of gastro-intestinal discomfort  
1874                          Reduction of gastric acid  

[1491 rows x 3 columns]

In [43]:
import snorkel
from snorkel.labeling import labeling_function
ABSTAIN = -1
MAINTENANCE = 0
ENHANCEMENT = 1
REDUCTION = 2
@labeling_function()
def lf_maint(x):
    return MAINTENANCE if "maintenance" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_maint2(x):
    return MAINTENANCE if "normal" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_enhan(x):
    return ENHANCEMENT if "enhance" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_enhan2(x):
    return ENHANCEMENT if "improves" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_enhan3(x):
    return ENHANCEMENT if "improvement" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_enhan4(x):
    return ENHANCEMENT if "acceleration" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_enhan5(x):
    return ENHANCEMENT if "increase" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_enhan6(x):
    return ENHANCEMENT if "increases" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_reduc(x):
    return REDUCTION if "reduction" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_reduc2(x):
    return REDUCTION if "reduces" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_reduc3(x):
    return REDUCTION if "lower" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_reduc4(x):
    return REDUCTION if "protection" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_reduc5(x):
    return REDUCTION if "protects" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_reduc6(x):
    return REDUCTION if "neutralisation" in x.text.lower() else ABSTAIN
@labeling_function()
def lf_reduc7(x):
    return REDUCTION if "neutralises" in x.text.lower() else ABSTAIN

In [44]:
from snorkel.labeling.model import LabelModel
from snorkel.labeling import PandasLFApplier

lfs = [lf_maint,lf_maint2, lf_enhan, lf_enhan2, lf_enhan3, lf_enhan4, lf_enhan5, lf_enhan6, lf_reduc, lf_reduc2, lf_reduc3, lf_reduc4, lf_reduc5, lf_reduc6, lf_reduc7]

applier = PandasLFApplier(lfs)
L_train = applier.apply(data)

label_model = LabelModel(cardinality=3, verbose=True)
label_model.fit(L_train, n_epochs=500, log_freq=50, seed=123)
data["label"] = label_model.predict(L=L_train, tie_break_policy="abstain")

100%|██████████| 500/500 [00:00<00:00, 853.75epoch/s]


In [47]:
data.groupby(["label"]).count()["text"]

label
-1    947
 0    205
 1    255
 2     84
Name: text, dtype: int64

In [20]:
data = data[data["label"] != -1]
data = data.reset_index()

In [21]:
data = data.drop(columns=["index"])

In [22]:
for i in range(len(data["text"])):
  print(i)
  print(data["text"].iloc[i])

0
"Traditionally used to enhance the renal elimination of water" / "Traditionally used to facilitate the elimination functions of the organism" / "Used to enhance the renal elimination of water" / "Used to facilitate the elimination functions of the organism" / "Used for the drainage of the organism" / "Diuretic effect" / "Stimulates the elimination function of the organism" / "Contributes to the functionary of the urinary tract"
1
"Used to enhance the renal elimination of water" / "Used to facilitate the elimination functions of the organism" / "Used for the drainage of the organism" / "Stimulates the elimination function of the organism" / "Contributes to the functionary of the urinary tract"
2
"Used to enhance the renal elimination of water" / "Used to facilitate the elimination functions of the organism" / "Used for the drainage of the organism" / "Stimulates the elimination function of the organism" / "Contributes to the functionary of the urinary tract"
3
"Used to facilitate the 

In [23]:
#Remove entries in other languages
data = data.drop([100,101,102,103,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,381,382,382,383,384,385,386,387,388,289,390,537])


In [24]:
data.groupby("label").count()["text"]

label
0    177
1    255
2     82
Name: text, dtype: int64

From the cell above, we can see that this labeling approach resulted in 177 entries labeled as "Maintenance of a function", 255 entries labeled as "Enhancing a function" and 82 entries labeled as "Reducing a risk factor". This brings the total of newly labeled entries to 544 entries, which corresponds to a 218,72% increase in data size

In [25]:
#Create csv file from new database
data.to_csv('/content/newdata.csv', index=None)